test: Support multiple runs for Pipeline run tests

silvanocerza opened the pull request:
This PR changes a bit the boilerplate functions that handle running the `Pipeline` in the `run` tests.

While I was migrating the other tests I noticed that some tests run a `Pipeline` multiple times, mostly when it has conditional branching, loops, and the likes. I thought of just running a `Pipeline` once but some actually require to be run with different inputs to actually verify their correct behaviour.

I tried using `@pytest.mark.parametrize` on the `@given` steps that define the `Pipeline` to test but that doesn't work, `pytest-bdd` forbids that.

Another option would have been to define the function that returns the `Pipeline` multiple times as different `@given` steps, one for each different input. That would be more error prone in my opinion as the `Pipeline` would be defined multiple times.

Best solution would be defining these parameters in the gherkin file, but that's not possible as that would have to be Python code. That would complicate things too much as we'd have to `eval`uate parts of the steps, making the gherkin less readable, more frustrating to work with, and ton of othe problems.

This is the simplest and fastest solution and is good enough for the time being .

masci commented on test/core/pipeline/features/README.md:
````diff
@@ -49,6 +49,28 @@ def pipeline_that_is_linear():
     )
 ```
 
+Some kind of `Pipeline`s requires multiple runs to correctly verify their correct working, e.g those with multiple branches.
````
```suggestion
Some kinds of `Pipeline`s require multiple runs to verify they work correctly, for example those with multiple branches.
```

masci commented on test/core/pipeline/features/README.md:
````diff
@@ -49,6 +49,28 @@ def pipeline_that_is_linear():
 ```
 
+Some kind of `Pipeline`s requires multiple runs to correctly verify their correct working, e.g those with multiple branches.
+Because of this we also support functions that return lists of inputs, expected outputs and expected run orders.
````
```suggestion
For this reason we also support functions returning a "list of inputs", a "list of expected outputs" and a "list of expected run orders" (all the lists have the same size).
```

masci commented on test/core/pipeline/features/README.md:
```diff
@@ -49,6 +49,28 @@ def pipeline_that_is_linear():
 
+Some kind of `Pipeline`s requires multiple runs to correctly verify their correct working, e.g those with multiple branches.
+Because of this we also support functions that return lists of inputs, expected outputs and expected run orders.
+If we would have needed to run the above `Pipeline` multiple times we could do something like this:
```
```suggestion
For example, we could test two different runs of the same pipeline like this:
```

silvanocerza commented:
Merging even with failing tests as they're failing for unrelated reasons that we just fixed. 
We just changed the README in that commit, and I don't want to wait for tests to run again.
